man/systemd-stub.xml

```diff
@@ -77,7 +77,17 @@
       <listitem><para>A <literal>.splash</literal> section with an image (in the Windows
       <filename>.BMP</filename> format) to show on screen before invoking the kernel.</para></listitem>
 
-      <listitem><para>A <literal>.dtb</literal> section with a compiled binary DeviceTree.</para></listitem>
+      <listitem><para>Zero or more <literal>.dtb</literal> sections. If only one such section is present, it
+      will always be passed to the kernel. If there at least two of them, stub will first try to match the
```

Comment:
so, one thing i am wondering about: what if i want to build a UKI that shall override the DTB of exactly one device, but on all other devices use the fw provided DTB. With this current logic I couldn't do this.

hence, my question: shouldn't we use a slightly different section name if want the matching thing to be done? i.e. `.dtbmatch` or so?

Then, `.dtb` would be a singleton as always. But if you do `.dtbmatch` instead, it would be a matching thing, and you could have one or more of them.

Reply:
it could even be used in conjunction then. i.e. we could say if a `.dtbmatch` is found that matches, uses it. If not, use `.dtb` if it exists. If neither, use fw provided dtb.

Reply:
> Then, .dtb would be a singleton as always

I believe it already isn't according to the spec. It's already allowed to have more than one `.dtb`:

> Optionally, one or more compiled Device Trees, for systems which need it, each in its separate .dtb section. If multiple .dtb sections exist then one of them is selected according to an implementation-specific algorithm.

Reply:
> so, one thing i am wondering about: what if i want to build a UKI that shall override the DTB of exactly one device, but on all other devices use the fw provided DTB. With this current logic I couldn't do this.

You can add a dummy `.dtb` which will never match and it should work exactly as described. This is a bit hacky, but at least we're not introducing new sections (BTW, `.dtbmatch` is too long) with basically the same content but slightly different semantics. Like we already ha

Reply:
> I believe it already isn't according to the spec. It's already allowed to have more than one .dtb:

Well, this is underspecified anyway, we can drop it again from the spec. And the new behaviour woudl be kinda compatible with the existing spec: our implementatoin defined algorithm would be to just pick he first .dtb as fallback, and of story.

> You can add a dummy .dtb which will never match and it should work exactly as described. This is a bit hacky, but at least we're not introducing new sections (BTW, .dtbmatch is too long) with basically the same content but slightly different semantics. Like we already ha

What's so bad about adding another section type? labels are free.

Reply:
Apart from Occam's razor, it's a bit confusing: why have 2 section type which do basically the same?
It'll be a bit painful to test (because UEFI is apparently broken with sd-boot on my device) and will also more changes across the whole codebase without actually introducing any features. Do you know any other cases where it would be crucial to have a separate section type? 
And also name needs to be different, maybe `.dtbauto` or something else?

Either way, I don't have a strong opinion here, but I'd like to hear a third opinion.

Reply:
> why have 2 section type which do basically the same?

but they don't. one does matching, the other one doesn't. i find that quite a fundemental differently.

The hence my suggestion to name this `.dtbmtch` or so, to emphasize exactly this fact. `.dtb` would be dumb, it would apply regardless of any matching. `.dtbmtch` would enable this special matching logic. It's very easy to explain that way, in particular if you combine them: if we find a matching `.dtbmtch`, we use it, otherwise we use the dumb `.dtb` regarldess if it matches or not.

Reply:
(i don't care too much about the name. .dtbauto sounds fine to me too)

Reply:
Right now, I am not aware that anything in systemd (or outside) actually supported multiple .dtb properly. With this we'll add tis after all, and I think that gives us the liberty to tweak the spec a bit.

Reply:
Done